### ISSUE TITLE
Hessian (w.r.t inputs) calculation in PyTorch differs from FuncTorch

### ISSUE THREAD

**Issue by AlphaBetaGamma96** [pytorch/functorch#977]
Hi All,

I've been trying to calculate the Hessian of the output of my network with respect to its inputs within FuncTorch. I had a version within PyTorch that supports batches, however, they seem to disagree with each other and I have no idea why they don't give the same results. Something is clearly wrong, I know my PyTorch version is right so either there's an issue in my version of FuncTorch or I've implemented it wrong in FuncTorch. 

Also, how can I use the `has_aux` flag in `jacrev` to return the jacobian from the first `jacrev` so I don't have to repeat the jacobian calculation?

The only problem with my example is that it uses `torch.linalg.slogdet` and from what I remember FuncTorch can't vmap over `.item()`. I do have my own fork of pytorch where I edited the backward to remove the `.item()` call so it works with vmap. Although, it's not the greatest implementation as I just set it to the default `nonsingular_case_backward` like so,
```
Tensor slogdet_backward(const Tensor& grad_logabsdet,
                        const Tensor& self,
                        const Tensor& signdet, const Tensor& logabsdet) {
  auto singular_case_backward = [&](const Tensor& grad_logabsdet, const Tensor& self) -> Tensor {
    Tensor u, sigma, vh;
    std::tie(u, sigma, vh) = at::linalg_svd(self, false);
    Tensor v = vh.mH();
    // sigma has all non-negative entries (also with at least one zero entry)
    // so logabsdet = \sum log(abs(sigma))
    // but det = 0, so backward logabsdet = \sum log(sigma)
    auto gsigma = grad_logabsdet.unsqueeze(-1).div(sigma);
    return svd_backward({}, gsigma, {}, u, sigma, vh);
  };

  auto nonsingular_case_backward = [&](const Tensor& grad_logabsdet, const Tensor& self) -> Tensor {
    // TODO: replace self.inverse with linalg_inverse
    return unsqueeze_multiple(grad_logabsdet, {-1, -2}, self.dim()) * self.inverse().mH();
  };

  auto nonsingular = nonsingular_case_backward(grad_logabsdet, self);
  return nonsingular;
}
```

My 'minimal' reproducible script is below with the output shown below that. It computes the Laplacian via a PyTorch method and via FuncTorch for a single sample of size `[A,1]` where `A` is the number of input nodes to the network.
```
import torch
import torch.nn as nn
from torch import Tensor
import functorch
from functorch import jacrev, jacfwd, hessian, make_functional, vmap
import time 

_ = torch.manual_seed(0)

print("PyTorch version:   ", torch.__version__)
print("CUDA version:      ", torch.version.cuda)
print("FuncTorch version: ", functorch.__version__)

def sync_time() -> float:
  torch.cuda.synchronize()
  return time.perf_counter()

B=1 #batch
A=3 #input nodes

device=torch.device("cuda")

class model(nn.Module):

  def __init__(self, num_inputs, num_hidden):
    super(model, self).__init__()
    
    self.num_inputs=num_inputs
    self.func = nn.Tanh()
    
    self.fc1 = nn.Linear(2, num_hidden)
    self.fc2 = nn.Linear(num_hidden, num_inputs)
  
  def forward(self, x):
    """
    Takes x in [B,A,1] and maps it to sign/logabsdet value in Tuple([B,], [B,])
    """
    
    idx=len(x.shape)
    rep=[1 for _ in range(idx)]
    rep[-2] = self.num_inputs
    g = x.mean(dim=(idx-2), keepdim=True).repeat(*rep)
    f = torch.cat((x,g), dim=-1)

    h = self.func(self.fc1(f))
    
    mat = self.fc2(h)
    sgn, logabs = torch.linalg.slogdet(mat)
    return sgn, logabs

net = model(A, 64)
net = net.to(device)

fnet, params = make_functional(net)

def logabs(params, x):
  _, logabs = fnet(params, x)
  #print("functorch logabs: ",logabs)
  return logabs


def kinetic_pytorch(xs: Tensor) -> Tensor:
  """Method to calculate the local kinetic energy values of a netork function, f, for samples, x.
  The values calculated here are 1/f d2f/dx2 which is equivalent to d2log(|f|)/dx2 + (dlog(|f|)/dx)^2
  within the log-domain (rather than the linear-domain).

  :param xs: The input positions of the many-body particles
  :type xs: class: `torch.Tensor`
  """
  xis = [xi.requires_grad_() for xi in xs.flatten(start_dim=1).t()]
  xs_flat = torch.stack(xis, dim=1)

  _, ys = net(xs_flat.view_as(xs))
  #print("pytorch logabs: ",ys)
  ones = torch.ones_like(ys)

  #df_dx calculation
  (dy_dxs, ) = torch.autograd.grad(ys, xs_flat, ones, retain_graph=True, create_graph=True)


  #d2f_dx2 calculation (diagonal only)
  lay_ys = sum(torch.autograd.grad(dy_dxi, xi, ones, retain_graph=True, create_graph=False)[0] \
                for xi, dy_dxi in zip(xis, (dy_dxs[..., i] for i in range(len(xis))))
  )
  #print("(PyTorch): ",lay_ys, dy_dxs)
  
  ek_local_per_walker = -0.5 * (lay_ys + dy_dxs.pow(2).sum(-1)) #move const out of loop?
  return ek_local_per_walker
  
jacjaclogabs = jacrev(jacrev(logabs, argnums=1), argnums=1)
jaclogabs = jacrev(logabs, argnums=1)
  
def kinetic_functorch(params, x):
  d2f_dx2 = vmap(jacjaclogabs, in_dims=(None, 0))(params, x)
  df_dx = vmap(jaclogabs, in_dims=(None, 0))(params, x)
  #print("(FuncTorch): ", d2f_dx2.squeeze(-3).squeeze(-1).diagonal(-2,-1).sum(-1), df_dx)
  #remove the trailing 1's so it's an A by A matrix 
  return -0.5 * d2f_dx2.squeeze(-3).squeeze(-1).diagonal(-2,-1).sum(-1) + df_dx.squeeze(-1).pow(2).sum(-1)

x = torch.randn(B,A,1,device=device) #input Tensor 

print("\nd2f/dx2, df/dx: ")
t1=sync_time()
kin_pt = kinetic_pytorch(x)
t2=sync_time()
t3=sync_time()
kin_ft = kinetic_functorch(params, x)
t4=sync_time()

print("\nWalltime: ")
print("PyTorch:   ",t2-t1)
print("FuncTorch: ",t4-t3, "\n")

print("Results: ")
print("PyTorch: ",kin_pt)
print("FuncTorch: ",kin_ft)
```

This script returns 
```
PyTorch version:    1.12.0a0+git7c2103a
CUDA version:       11.6
FuncTorch version:  0.2.0a0+9d6ee76

d2f/dx2, df/dx: 

Walltime: 
PyTorch:    0.4822753759999614
FuncTorch:  0.004898710998531897 

Results: 
PyTorch:  tensor([1.3737], device='cuda:0', grad_fn=<MulBackward0>)    # should be the same values
FuncTorch:  tensor([7.8411], device='cuda:0', grad_fn=<AddBackward0>) # the jacobian matches, but hessian does not
```
Thanks for the help in advance! :)

**Comment by AlphaBetaGamma96** [comment]
Ah, so I managed to find the error in so I was missing an argument in `.diagonal` it was `.diagonal(-2,-1)` but it should've been `.diagonal(0,-2,-1)`, and I was missing a bracket. And with these changes PyTorch and FuncTorch match so it's all good.

However, I also seek to take per-sample gradient for this terms as well, and I get a same mismatch between PyTorch's reverse-mode AD and FuncTorch per-sample gradients.

```
import torch
import torch.nn as nn
from torch import Tensor
import functorch
from functorch import jacrev, jacfwd, hessian, make_functional, vmap, grad
import time 
from copy import deepcopy

_ = torch.manual_seed(0)

#version info
print("PyTorch version:   ", torch.__version__)
print("CUDA version:      ", torch.version.cuda)
print("FuncTorch version: ", functorch.__version__)

def sync_time() -> float:
  torch.cuda.synchronize()
  return time.perf_counter()

B=1  #batch
A=3  #input nodes
H=16 #number of hidden nodes

device=torch.device("cuda")

class model(nn.Module):

  def __init__(self, num_inputs, num_hidden):
    super(model, self).__init__()
    
    self.num_inputs=num_inputs
    self.func = nn.Tanh()
    
    self.fc1 = nn.Linear(2, num_hidden)
    self.fc2 = nn.Linear(num_hidden, num_inputs)
  
  def forward(self, x):
    """
    Takes x in [B,A,1] and maps it to sign/logabsdet value in Tuple([B,], [B,])
    """
    
    idx=len(x.shape)
    rep=[1 for _ in range(idx)]
    rep[-2] = self.num_inputs
    g = x.mean(dim=(idx-2), keepdim=True).repeat(*rep)
    f = torch.cat((x,g), dim=-1)

    h = self.func(self.fc1(f))
    
    mat = self.fc2(h)
    sgn, logabs = torch.linalg.slogdet(mat)
    return sgn, logabs

net = model(A, H)
net = net.to(device)

x = torch.randn(B,A,1,device=device) #input data
fnet, params = make_functional(net)

def logabs(params, x):
  _, logabs = fnet(params, x)
  return logabs

def kinetic_pytorch(xs: Tensor) -> Tensor:
  """Method to calculate the local kinetic energy values of a netork function, f, for samples, x.
  The values calculated here are 1/f d2f/dx2 which is equivalent to d2log(|f|)/dx2 + (dlog(|f|)/dx)^2
  within the log-domain (rather than the linear-domain).

  :param xs: The input positions of the many-body particles
  :type xs: class: `torch.Tensor`
  """
  xis = [xi.requires_grad_() for xi in xs.flatten(start_dim=1).t()]
  xs_flat = torch.stack(xis, dim=1)

  _, ys = net(xs_flat.view_as(xs))
  #print("pytorch logabs: ",ys)
  ones = torch.ones_like(ys)

  #df_dx calculation
  (dy_dxs, ) = torch.autograd.grad(ys, xs_flat, ones, retain_graph=True, create_graph=True)


  #d2f_dx2 calculation
  lay_ys = sum(torch.autograd.grad(dy_dxi, xi, ones, retain_graph=True, create_graph=False)[0] \
                for xi, dy_dxi in zip(xis, (dy_dxs[..., i] for i in range(len(xis))))
  )
  
  ek_local_per_walker = -0.5 * (lay_ys + dy_dxs.pow(2).sum(-1)) #move const out of loop?
  return ek_local_per_walker
  

def kinetic_functorch(params, X):
  calc_jacobian = jacrev(logabs, argnums=1) #do once, and re-use?
  calc_hessian = jacrev(jacrev(logabs, argnums=1), argnums=1)

  return -0.5*torch.sum(calc_hessian(params, X).squeeze(-3).squeeze(-1).diagonal(0,-2,-1) + calc_jacobian(params, X).squeeze(-1).pow(2), dim=-1)

#per-sample gradients for local energy w.r.t params via FuncTorch
elocal_grad_ft = vmap(grad(kinetic_functorch, argnums=0), in_dims=(None, 0))(params, x)
elocal_grad_ft = [p.clone().mean(dim=0) for p in elocal_grad_ft]

#PyTorch gradients via reverse-mode AD
net.zero_grad()
kin_pt = kinetic_pytorch(x)
loss=torch.mean(kin_pt)
loss.backward()
elocal_grad_pt = [param.grad for param in net.parameters()]

for i, (grad_ft, grad_pt) in enumerate(zip(elocal_grad_ft, elocal_grad_pt)):
  print("Param group: %i Equal?: %s" % (i, torch.allclose(grad_ft, grad_pt)))
  #print(i, grad_ft.flatten(), grad_pt.flatten())

#done
```

This scripts returns 
```
PyTorch version:    1.12.0a0+git7c2103a
CUDA version:       11.6
FuncTorch version:  0.2.0a0+9d6ee76
Param group: 0 Equal?: False
Param group: 1 Equal?: False
Param group: 2 Equal?: False
Param group: 3 Equal?: False
```

Perhaps I might be misunderstanding how reverse-mode AD compares with per-sample gradient with respect to FuncTorch. But I'm pretty sure taking the mean over the per-sample gradients should equal the reverse-mode AD (potentially times by the batch size). The example script above runs for a batch size of 1, so that _shouldn't_ be an issue. 

My Fork of PyTorch can be found here: https://github.com/AlphaBetaGamma96/pytorch/blob/custom_slogdet_backward_attempt/torch/csrc/autograd/FunctionsManual.cpp

TL;DR the mean of my per-sample gradients of the Hessian (with respect to the inputs) of my network do not match with reverse-mode AD results. I'm most likely doing something wrong but I can't see it at all as I'm pretty sure they should equal! 

Thanks!

**Comment by zou3519** [comment]
> The only problem with my example is that it uses torch.linalg.slogdet and from what I remember FuncTorch can't vmap over .item(). I do have my own fork of pytorch where I edited the backward to remove the .item() call so it works with vmap.

PyTorch actually changed recently to not have a .item() call in slogdet backward, so your above script (https://gist.github.com/zou3519/4e0b976fc32f3e989592dab2da90ce89) runs on the latest pytorch nightly + the latest functorch.

That being said, I am getting the same output as you, so I am going to dig into it

**Comment by zou3519** [comment]
Running your new code, it looks like kinetic_functorch and kinetic_pytorch return different things. Is this intentional?
```
[ins] In [16]: energy_ft = kinetic_functorch(params, x)
          ...: energy_pt = kinetic_pytorch(x)
/raid/rzou/pt/binaries4-env/lib/python3.10/site-packages/torch/autograd/__init__.py:294: UserWarning: There is a performance drop
 because we have not yet implemented the batching rule for aten::linalg_solve. Please file us an issue on GitHub so that we can p
rioritize its implementation. (Triggered internally at /private/home/rzou/functorch4/functorch/csrc/BatchedFallback.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass

[ins] In [17]: energy_ft
Out[17]:
tensor([[[ 3.3275, -9.5351, -4.1663]]], device='cuda:0',
       grad_fn=<MulBackward0>)

[ins] In [18]: energy_pt
Out[18]: tensor([1.2722], device='cuda:0', grad_fn=<MulBackward0>)
```

**Comment by AlphaBetaGamma96** [comment]
> Running your new code, it looks like kinetic_functorch and kinetic_pytorch return different things. Is this intentional?

No not all, they should return the same quantity which is the gradient of the kinetic (lapalacian) value w.r.t params. It's odd they return different shapes as well. Just to clarify, is this the code for the first comment I made or the second? The first one had a missing bracket, and pytorch matched functorch when computing the laplacian (from the log-domain). 

However, when doing the per-sample gradients with respect to params (2nd comment). They don't match. I got the same shapes but different values, and it seemed to emerge from the Hessian part of the computation.

**Comment by AlphaBetaGamma96** [comment]
Ok, I'll quickly re-run this and check it in case I've missed anything. I assume I should download the most recent versions of pytorch and functorch? As the slogdet_backward issue is solved, I can just use the getting started page to download rather than from source, right?

**Comment by AlphaBetaGamma96** [comment]
Ok, so installed the nightly (and functorch from source). I could reproduce your error, if the `kinetic_functorch` is vmap'd then the outputs match.
```
energy_ft = vmap(kinetic_functorch, in_dims=(None, 0))(params, x)
energy_pt = kinetic_pytorch(x)
```
returns
```
tensor([1.2722], device='cuda:0', grad_fn=<MulBackward0>)
tensor([1.2722], device='cuda:0', grad_fn=<MulBackward0>)
```

However, the per-sample gradients still differ. 

I've made the changes (with its output) in a gist here-> https://gist.github.com/AlphaBetaGamma96/6b6ff78138196ad24c11ae34b8f2380e

**Comment by zou3519** [comment]
Thanks for the clarification. Let me dig into that some more

**Comment by zou3519** [comment]
When you say you want per-sample-gradient, what exactly do you mean? The input x has shape `[B, A, 1]` -- do you want B different gradients, or A different gradients?

**Comment by zou3519** [comment]
I figured out what the difference is. In your code for `kinetic_pytorch`, you have `create_graph=False`. Is this intentional?


<img width="888" alt="image" src="https://user-images.githubusercontent.com/5652049/181308406-d1b25eaa-278b-4dd6-938f-c04e861e4763.png">

**Comment by zou3519** [comment]
(Sorry, I hit the wrong button). The effect of the create_graph=False is that when you're computing a gradient over the above quantity using regular PyTorch, pytorch does not actually backprop through your d2f_dx2 computation.

If the create_graph=False is indeed intentional, then it should be possible to rewrite the functorch part to match (I'll let you know how if this is what is desired).

**Comment by zou3519** [comment]
> Actually yes, the create_graph=False flag is the correct behaviour.

You probably want some `detach()` calls in the functorch version so that gradients don't flow through that part of the computation.

> Or is it the same numerical stability as pytorch?

Should be the same numerical stability. We leverage the same autograd formulas as regular PyTorch

**Comment by AlphaBetaGamma96** [comment]
> You probably want some detach() calls in the functorch version so that gradients don't flow through that part of the computation.

yes I do utilise a `.detach()` function elsewhere to stop backprop flowing through that in the `pytorch` version. 

Thanks @zou3519, and apologizes for missing the `create_graph=False` issue!

**Comment by AlphaBetaGamma96** [comment]
I will do, @zou3519! I was going to quickly ask if #984 has been resolved? As there were some [changes](https://github.com/pytorch/pytorch/commit/11a0b165da7f014c642f00efc68da4801a536cce#diff-cc14584b146892dbe216dc5d965a06149738a61611df0408ab61ec0b59981d01L784) made it seems that may have resolved my initial issue. Shall I upgrade and try?

**Comment by zou3519** [comment]
That's still in progress over at https://github.com/pytorch/pytorch/pull/82177
